File: python/cdp/analytics.py
import contextlib
import functools
import hashlib
import inspect
import json
import time
import traceback

import requests
from pydantic import BaseModel

# This is a public client id for the analytics service
public_client_id = "54f2ee2fb3d2b901a829940d70fbfc13"


class ErrorEventData(BaseModel):
    """The data in an error event."""

    api_key_id: str  # The API key ID
    method: str  # The API method where the error occurred, e.g. createAccount, getAccount
    message: str  # The error message
    name: str  # The name of the event. This should match the name in AEC
    stack: str | None = None  # The error stack trace


EventData = ErrorEventData


async def send_event(event: EventData) -> None:
    """Send an analytics event to the default endpoint.

    Args:
        event: The event data containing event-specific fields

    Returns:
        None - resolves when the event is sent

    """
    timestamp = int(time.time() * 1000)

    enhanced_event = {
        "user_id": event.api_key_id,
        "event_type": event.name,
        "platform": "server",
        "timestamp": timestamp,
        "event_properties": {
            "project_name": "cdp-sdk",
            "cdp_sdk_language": "python",
            **event.model_dump(),
        },
    }

    events = [enhanced_event]
    stringified_event_data = json.dumps(events)
    upload_time = str(timestamp)

    checksum = hashlib.md5((stringified_event_data + upload_time).encode("utf-8")).hexdigest()

    analytics_service_data = {
        "client": public_client_id,
        "e": stringified_event_data,
        "checksum": checksum,
    }

    api_endpoint = "https://cca-lite.coinbase.com"
    event_path = "/amp"
    event_endpoint = f"{api_endpoint}{event_path}"

    requests.post(
        event_endpoint,
        headers={"Content-Type": "application/json"},
        json=analytics_service_data,
    )


def wrap_with_error_tracking(func, api_key_id: str):
    """Wrap a method with error tracking.

    Args:
        func: The function to wrap.
        api_key_id: The API key ID

    Returns:
        The wrapped function.

    """

    @functools.wraps(func)
    async def wrapper(*args, **kwargs):
        try:
            return await func(*args, **kwargs)
        except Exception as error:
            event_data = ErrorEventData(
                api_key_id=api_key_id,
                method=func.__name__,
                message=str(error),
                stack=traceback.format_exc(),
                name="error",
            )

            with contextlib.suppress(Exception):
                await send_event(event_data)

            raise error

    return wrapper


def wrap_class_with_error_tracking(cls, api_key_id: str):
    """Wrap all methods of a class with error tracking.

    Args:
        cls: The class to wrap.
        api_key_id: The API key ID

    Returns:
        The class with wrapped methods.

    """
    for name, method in inspect.getmembers(cls, inspect.isfunction):
        if not name.startswith("__"):
            setattr(cls, name, wrap_with_error_tracking(method, api_key_id))
    return cls

File: typescript/src/analytics.ts
import md5 from "md5";

/**
 * The data in an error event
 */
type ErrorEventData = {
  /**
   * The API method where the error occurred, e.g. createAccount, getAccount
   */
  method: string;
  /**
   * The error message
   */
  message: string;
  /**
   * The error stack trace
   */
  stack?: string;
  /**
   * The name of the event. This should match the name in AEC
   */
  name: "error";
};

type EventData = ErrorEventData & {
  apiKeyId: string;
};

// This is a public client id for the analytics service
const publicClientId = "54f2ee2fb3d2b901a829940d70fbfc13";

/**
 * Sends an analytics event to the default endpoint
 *
 * @param event - The event data containing event-specific fields
 * @returns Promise that resolves when the event is sent
 */
export async function sendEvent(event: EventData): Promise<void> {
  const timestamp = Date.now();

  const enhancedEvent = {
    user_id: event.apiKeyId,
    event_type: event.name,
    platform: "server",
    timestamp,
    event_properties: {
      project_name: "cdp-sdk",
      cdp_sdk_language: "typescript",
      ...event,
    },
  };

  const events = [enhancedEvent];
  const stringifiedEventData = JSON.stringify(events);
  const uploadTime = timestamp.toString();

  const checksum = md5(stringifiedEventData + uploadTime);

  const analyticsServiceData = {
    client: publicClientId,
    e: stringifiedEventData,
    checksum,
  };

  const apiEndpoint = "https://cca-lite.coinbase.com";
  const eventPath = "/amp";
  const eventEndPoint = `${apiEndpoint}${eventPath}`;

  await fetch(eventEndPoint, {
    method: "POST",
    mode: "no-cors",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(analyticsServiceData),
  });
}

/**
 * Wraps all methods of a class with error tracking.
 *
 * @param ClassToWrap - The class whose prototype methods should be wrapped.
 * @param apiKeyId - The API key ID to use for the error tracking.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function wrapClassWithErrorTracking(ClassToWrap: any, apiKeyId: string): void {
  const methods = Object.getOwnPropertyNames(ClassToWrap.prototype).filter(
    name => name !== "constructor" && typeof ClassToWrap.prototype[name] === "function",
  );

  for (const method of methods) {
    const originalMethod = ClassToWrap.prototype[method];
    ClassToWrap.prototype[method] = async function (...args: unknown[]) {
      try {
        return await originalMethod.apply(this, args);
      } catch (error) {
        if (!(error instanceof Error)) {
          return;
        }

        const { message, stack } = error;

        sendEvent({
          apiKeyId,
          method,
          message,
          stack,
          name: "error",
        }).catch(() => {
          // ignore error
        });

        throw error;
      }
    };
  }
}
